Add logical_xor operator

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24379 Implement operator logical_and
* #23978 Let logical_xor support non-bool tensors.
* #23916 Let logical_not support non-bool tensors.
* #23860 Recommend logical_not() instead of bitwise_not() when applying sub and neg on bool tensors.
* **#23847 Add logical_xor operator**
* #23839 Add logical_not operator.

Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
[WIP] Add logical_xor

Close #23836

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

gchanan reviewed (commented):
same comments as the logical_not PR; mainly return bool by default.

gchanan reviewed (changes requested):
.

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

gchanan commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -197,5 +198,24 @@ Tensor rsub(const Tensor& self, Scalar other, Scalar alpha) {
 
+Tensor& logical_xor_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(self.scalar_type() == kBool && other.scalar_type() == kBool,
+              "logical_xor only supports bool tensors.");
```
nit: currently only supports.

gchanan commented on test/test_torch.py:
```diff
@@ -1837,6 +1837,25 @@ def _test_logical_not(self, device):
+            self.assertEqual(expected_res, c)
+            # in-place
+            a.logical_xor_(b)
+            self.assertEqual(expected_res, a)
```
same as logical_not -- can we test `out=` with the wrong dtype?  (Also what happens?)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

gchanan commented:
clang tidy failure looks bogus.

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)

xuhdev committed:
Update on "Add logical_xor operator"


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23847

Differential Revision: [D16678300](https://our.internmc.facebook.com/intern/diff/D16678300)
